dlt/destinations/impl/sink/sink.py

```diff
@@ -0,0 +1,175 @@
+    destination_state,
+    commit_load_package_state,
+)
+
+from dlt.common.schema import Schema, TTableSchema, TSchemaTables
+from dlt.common.schema.typing import TTableSchema
+from dlt.common.storages import FileStorage
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.destination.reference import (
+    TLoadJobState,
+    LoadJob,
+    JobClientBase,
+)
+
+from dlt.destinations.impl.sink import capabilities
+from dlt.destinations.impl.sink.configuration import SinkClientConfiguration, TSinkCallable
+
+
+class SinkLoadJob(LoadJob, ABC):
+    def __init__(
+        self,
+        table: TTableSchema,
+        file_path: str,
+        config: SinkClientConfiguration,
+        schema: Schema,
+        destination_state: Dict[str, int],
+        resolved_callable: TSinkCallable,
+    ) -> None:
+        super().__init__(FileStorage.get_file_name_from_file_path(file_path))
+        self._file_path = file_path
+        self._config = config
+        self._table = table
+        self._schema = schema
+        self._resolved_callable = resolved_callable
+
+        self._state: TLoadJobState = "running"
+        self._storage_id = f"{self._parsed_file_name.table_name}.{self._parsed_file_name.file_id}"
+        try:
+            if self._config.batch_size == 0:
+                # on batch size zero we only call the callable with the filename
+                self.call_callable_with_items(self._file_path)
+            else:
+                current_index = destination_state.get(self._storage_id, 0)
+                for batch in self.run(current_index):
+                    self.call_callable_with_items(batch)
+                    current_index += len(batch)
+                    destination_state[self._storage_id] = current_index
+
+            self._state = "completed"
+        except Exception as e:
+            self._state = "retry"
+            raise e
+        finally:
+            # save progress
+            commit_load_package_state()
+
+    @abstractmethod
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        pass
+
+    def call_callable_with_items(self, items: TDataItems) -> None:
+        if not items:
+            return
+        # call callable
+        self._resolved_callable(items, self._table)
+
+    def state(self) -> TLoadJobState:
+        return self._state
+
+    def exception(self) -> str:
+        raise NotImplementedError()
+
+
+class SinkParquetLoadJob(SinkLoadJob):
+    def run(self, start_index: int) -> Iterable[TDataItems]:
+        # stream items
+        from dlt.common.libs.pyarrow import pyarrow
+
+        # guard against changed batch size after restart of loadjob
+        assert (
+            start_index % self._config.batch_size
+        ) == 0, "Batch size was changed during processing of one load package"
+
+        start_batch = start_index / self._config.batch_size
+        with pyarrow.parquet.ParquetFile(self._file_path) as reader:
```

Comment:
mhhhh this looks to me like `file_readers` - opposite to our `file_writers`. so you can read parquet and jsonl (with pua) / json with them.

makes sense to extract it now?

Reply:
Let's do a separate PR for this where filetypes are fully pluggable..